File: CPP/LL/02_add_2_no.cpp
#include<iostream>
using namespace std;

struct ListNode{
    int val;
    ListNode *next;
    ListNode (int x):val(x),next(nullptr){}
};

ListNode* addTwoNumber(ListNode * l1,ListNode * l2){
    ListNode * dummy =new ListNode(0);
    ListNode * curr=dummy;
    int carry=0;

    while(l1!=nullptr || l2!=nullptr || carry!=0){
        int sum=carry;
        if(l1!=nullptr){
            sum+=l1->val;
            l1=l1->next;
        }
        if(l2!=nullptr){
            sum+=l2->val;
            l2=l2->next;
        }
        carry=sum/10;
        curr->next=new ListNode(sum%10);
        curr=curr->next;
    }
    return dummy->next;
}
int main(){

    int n1,n2;
    cin>>n1;

    ListNode * l1=new ListNode(0);
    ListNode *curr=l1;

    for(int i=0;i<n1;i++){
        int x;
        cin>>x;
        curr->next=new ListNode(x);
        curr=curr->next;
    }

    cin>>n2;

    ListNode * l2=new ListNode(0);
    curr=l2;

    for(int i=0;i<n2;i++){
        int x;
        cin>>x;
        curr->next=new ListNode(x);
        curr=curr->next;
    }
    ListNode * res=addTwoNumber(l1->next,l2->next);
    while(res!=nullptr){
        cout<<res->val<<" ";
        res=res->next;
    }
    return 0;
    

}

File: Python/02_Add_2_no.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1, l2):

        dummy = ListNode(0)
        current = dummy
        carry = 0

        while l1 is not None or l2 is not None or carry != 0:

            sum = carry

            if l1 is not None:
                sum += l1.val
                l1 = l1.next

            if l2 is not None:
                sum += l2.val
                l2 = l2.next

            carry = sum // 10

            current.next = ListNode(sum % 10)
            current = current.next

        return dummy.next


# Create linked list
def create_linked_list(arr):
    dummy = ListNode(0)
    current = dummy

    for value in arr:
        current.next = ListNode(value)
        current = current.next

    return dummy.next


# Print linked list
def print_linked_list(head):
    result = []

    while head is not None:
        result.append(head.val)
        head = head.next

    print(result)


# -------------------------
# Take input from user
# -------------------------

l1_input = input("Enter elements of l1 separated by space: ")
l2_input = input("Enter elements of l2 separated by space: ")

l1 = create_linked_list(map(int, l1_input.split()))
l2 = create_linked_list(map(int, l2_input.split()))


# Solve
solution = Solution()
answer = solution.addTwoNumbers(l1, l2)


# Print answer
print("Result:", end=" ")
print_linked_list(answer)